system-addon/lib/TelemetrySender.jsm

```
@@ -0,0 +1,126 @@
+  this._onEndpointPrefChange = this._onEndpointPrefChange.bind(this);
+  this._prefs.observe(ENDPOINT_PREF, this._onEndpointPrefChange);
+
+  if (this.enabled) {
+    Services.obs.addObserver(this, COMPLETE_NOTIF, true);
+    Services.obs.addObserver(this, ACTION_NOTIF, true);
+    Services.obs.addObserver(this, PERFORMANCE_NOTIF, true);
+    Services.obs.addObserver(this, UNDESIRED_NOTIF, true);
+  }
+}
+
+TelemetrySender.prototype = {
+  /* We need this to be multi-process safe while using the observer service
+   * (though it's not obvious to me how the weak reference helps that). */
+  QueryInterface: XPCOMUtils.generateQI([
+    Ci.nsIObserver,
+    Ci.nsISupportsWeakReference
+  ]),
+
+  observe(subject, topic, data) {
+    if (topic === COMPLETE_NOTIF || topic === ACTION_NOTIF || topic === PERFORMANCE_NOTIF || topic === UNDESIRED_NOTIF) {
+      this._sendPing(data);
+    }
+  },
+
+  _onEndpointPrefChange(prefVal) {
+    this._pingEndpoint = prefVal;
+  },
+
+  _onLoggingPrefChange(prefVal) {
+    this.logging = prefVal;
+  },
+
+  _onTelemetryPrefChange(prefVal) {
+    if (this.enabled && !prefVal) {
+      Services.obs.removeObserver(this, COMPLETE_NOTIF);
+      Services.obs.removeObserver(this, ACTION_NOTIF);
+      Services.obs.removeObserver(this, PERFORMANCE_NOTIF);
+      Services.obs.removeObserver(this, UNDESIRED_NOTIF);
+    } else if (!this.enabled && prefVal) {
+      Services.obs.addObserver(this, COMPLETE_NOTIF, true);
+      Services.obs.addObserver(this, ACTION_NOTIF, true);
+      Services.obs.addObserver(this, PERFORMANCE_NOTIF, true);
+      Services.obs.addObserver(this, UNDESIRED_NOTIF, true);
+    }
+
+    this.enabled = prefVal;
+  },
+
+  async _sendPing(data) {
+    if (this.logging) {
+      // performance related pings cause a lot of logging, so we mute them
+      if (JSON.parse(data).action !== "activity_stream_performance") {
+        Cu.reportError(`TELEMETRY PING: ${data}`);
```

Do we want to report an error here, or just log it to console?